tools/c_api/generate_matrix.py

```diff
@@ -72,6 +92,56 @@
 #endif
 \n''')
 
+# Write Options accessors
```

Comment:
I was trying to follow the style elsewhere in the function.  And I think properly indenting the generator source will result in weird indenting in the generated code or vis-versa (unless I'm misunderstanding python's rules for multi-line strings; either way the non-obviousness suggests it's use should be avoided.)

Also, note that we're interspersing writes the the header with writes to the cc file.  E.g.,
```
Write declaration to header
Write declaration to cc
Write { to cc
Write body to cc
Write } to cc
```

Reply:
See if [what I did](https://github.com/icl-utk-edu/slate/pull/152/files/dddc39e09486a927f0bb1a655a8ad31211cab626..202b09402baf7a19cf178ec15b8e61ee941e079f) looks okay. Yes, the strings inside the loop lack indenting, but the C++ code to be generated is much easier to follow.

Reply:
I think your right that it looks better.  I also like the use of format strings for doing the substitution.

It might be worth adding to #156 to extending that style to the rest of the generator files.

Reply:
Alternatively, the strings could be created before the loop and formatted inside the loop.
```python
#-------------------------------------------------------------------------------
void = "void"

hh = '''
{void:{width}} slate_Options_set_{opt_name}( slate_Options opts, {val_type} value );
{val_type:{width}} slate_Options_get_{opt_name}( slate_Options opts, {val_type} defval );
'''

cc = '''
void slate_Options_set_{opt_name}( slate_Options opts, {val_type} value )
{{
    auto& opts_ = *static_cast<slate::Options*>( opts );
    opts_[ slate::Option::{opt_name} ] = value;
}}

{val_type} slate_Options_get_{opt_name}( slate_Options opts, {val_type} defval )
{{
    auto& opts_ = *static_cast< slate::Options* >( opts );
    auto defval_ = static_cast< slate::OptValueType< slate::Option::{opt_name} >::T >( defval );
    return static_cast<{val_type}>( slate::get_option< slate::Option::{opt_name} >( opts_, defval_ ));
}}
'''

for opt_name, val_type in options:
    width = max( 4, len( val_type ) )  # max( "void", val_type )
    file_hh.write( hh.format( opt_name=opt_name, val_type=val_type, width=width, void=void ) )
    file_cc.write( cc.format( opt_name=opt_name, val_type=val_type ) )
# end for
```

Reply:
I feel like putting the strings before the loop is a touch less intuitive since we're "using" the variables before their defined.  But, I'd say it's a horse apiece.

Reply:
"a horse apiece"?

Yes, I wasn't too happy with putting strings before the loop.

Another alternative, inspired by how Python deals with docstrings by stripping a uniform amount of leading space from them.
```python
#-------------------------------------------------------------------------------
void = "void"
for opt_name, val_type in options:
    width = max( 4, len( val_type ) )  # max( "void", val_type )

    # trim strips a uniform amount of leading indent.
    file_hh.write( trim( f'''
        {void:{width}} slate_Options_set_{opt_name}( slate_Options opts, {val_type} value );
        {val_type:{width}} slate_Options_get_{opt_name}( slate_Options opts, {val_type} defval );
        ''' ) )

    file_cc.write( trim( f'''
        void slate_Options_set_{opt_name}( slate_Options opts, {val_type} value )
        {{
            auto& opts_ = *static_cast<slate::Options*>( opts );
            opts_[ slate::Option::{opt_name} ] = value;
        }}

        {val_type} slate_Options_get_{opt_name}( slate_Options opts, {val_type} defval )
        {{
            auto& opts_ = *static_cast< slate::Options* >( opts );
            auto defval_ = static_cast< slate::OptValueType< slate::Option::{opt_name} >::T >( defval );
            return static_cast<{val_type}>( slate::get_option< slate::Option::{opt_name} >( opts_, defval_ ));
        }}
        ''' ) )
# end for
```

Reply:
"It's a horse apiece" means the options are basically the same.  I hadn't realized that was a Minnesota-ism.

The third option looks good too, although I'm not sure where the trim function is from.  All I could find is [`textwrap.dedent'](https://docs.python.org/3/library/textwrap.html#textwrap.dedent)

Reply:
`trim` was a function I wrote, but `dedent` does more-or-less the same, so I'll use that. Thanks for finding it.

```
#-------------------------------------------------------------------------------
def trim( msg ):
    '''
    Removes a uniform amount of leading space from all lines of string msg,
    and returns the new string.
    '''
    s = re.search( r'^\n?( +)', msg )
    if (s):
        space = s.group( 1 )
        msg = re.sub( r'^' + space, r'', msg, 0, re.MULTILINE )
    return msg
# end
```

Reply:
I [updated the code](https://github.com/icl-utk-edu/slate/pull/152/files/dddc39e09486a927f0bb1a655a8ad31211cab626..a0777c0c54f902bfa6dba3648ff072ae564a597b) with the 2nd alternative, using `dedent`.